### ISSUE TITLE
Problem with multiple filtering  &  datetime filters

### ISSUE THREAD

**Issue by pabloitu** [SCECcode/pycsep#36]
Hello! By filtering a catalog using a list of filters (e.g catalog temporal filtering from the tutorial), the catalog is just filtered by last member in list. The problem can be reproduced as:

#### Input
```python3.7
import csep
import copy
from csep.utils.time_utils import strptime_to_utc_epoch,strptime_to_utc_datetime
from csep.core.catalogs import CSEPCatalog
# define dummy cat
date1 = strptime_to_utc_epoch('2009-01-01 00:00:00.0000')
date2 = strptime_to_utc_epoch('2010-01-01 00:00:00.0000')
date3 = strptime_to_utc_epoch('2011-01-01 00:00:00.0000')
Catalog = [(b'1', date1, 1.0, 1.0, 1.0, 1.0),
           (b'2', date2, 2.0, 2.0, 2.0, 2.0),
           (b'3', date3, 3.0, 3.0, 3.0, 3.0)]
test_cat1 = CSEPCatalog()
test_cat1.catalog = Catalog
test_cat2 = copy.deepcopy(test_cat1)
# Filter together
start_epoch = strptime_to_utc_epoch('2009-07-01 00:00:00.0')
end_epoch = strptime_to_utc_epoch('2010-07-01 00:00:00.0')
filters = [f'origin_time >= {start_epoch}', f'origin_time < {end_epoch}']    # should return only event 2
test_cat1.filter(filters)
print('Events after filter by list:', test_cat1.get_event_ids())
# Filter separately
for i in filters:
    test_cat2.filter(i)
print('Events after filter by two operators:' , test_cat2.get_event_ids(),'\n')
```
#### Output
```python3.7
Events after filter by list: [b'1' b'2']
Events after filter by two operators: [b'2'] 
```
I think is due to reassigning in each iteration of the filters list, the values of the original catalog object to the new filtered catalog. In this line:
https://github.com/SCECcode/pycsep/blob/17c40ccab99976f1d89252930d914cfa1b1349ba/csep/core/catalogs.py#L521
A solution would be maybe to replace line 521 with 522.

Also, the filter gives error with datetime objects:
#### Input
```python3.7
start_time = strptime_to_utc_datetime('2010-01-01 00:00:00.0000')
end_time = strptime_to_utc_datetime('2010-01-04 00:00:00.0000')
test_cat = csep.query_comcat(start_time, end_time, verbose=False)

# Filter together
start_epoch = strptime_to_utc_datetime('2010-01-02 00:00:00.0')
filter = f'origin_time >= {start_epoch}'
test_cat.filter(filter)
```
#### Output
```
 line 499, in filter
    name, oper, value = statements.split(' ')

ValueError: too many values to unpack (expected 3)
```
I believe is because f-string formatting a datetime object gives a space between date and time 
```
filter
'origin_time >= 2010-01-02 00:00:00+00:00'
```

Thank you!

**Comment by wsavran** [comment]
good catch on the first part! it looks like this line
```
filtered = self.catalog[operators[oper](self.catalog[name], float(value))]
```
was not supposed to be in the code. i am going to turn this example you wrote into a unit-test for this function. thank you for such a detailed error report. 

for the datetime problem, i think maybe we force the datetime string to have no spaces between the date and time. it would look something like this 
```
2010-01-02T00:00:00
```
do you have any ideas?

**Comment by pabloitu** [comment]
I believe when we f-string a datetime object, it is going to return always a date-time separated by space
for example, even if we assign a format separated by a T:
#### input
```
 a =datetime.datetime.strptime('2010-01-01T01:01:01.01', "%Y-%m-%dT%H:%M:%S.%f")
print(f'{a}')
```

#### output
```
'2010-01-01 01:01:01.010000'
```

So I guess, the only solution is to catch this in the catalog filter. Probably modifying this
https://github.com/SCECcode/pycsep/blob/d7d9c09f6906de22d3c46d79c54d3bfa9b45f398/csep/core/catalogs.py#L499
 
We can do a string split by one of the possible operators, but I guess we have to catch the variable to be filtered (e.g. when someone writes x>5 or 5<x).
  
```  
operators = {'>': operator.gt,
                     '<': operator.lt,
                     '>=': operator.ge,
                     '<=': operator.le,
                     '==': operator.eq}`
```

**Comment by wsavran** [comment]
i think what we are seeing the is __str__ method of the datetime object making it nice and human readable when passed to the print() function. 

```
>>> a = '2010-01-01T01:01:01.01'
>>> print(f'{a}')
2010-01-01T01:01:01.01
```

**Comment by pabloitu** [comment]
Yes exactly

**Comment by wsavran** [comment]
yeah, we do. 

fundamentally the function converts everything to a string, so we cant really get around this. i think a solution could be to add a try - catch to handle this case and then convert the datetime string to a datetime object. do you have any other ideas?

**Comment by wsavran** [comment]
ok i pushed another commit to this pull request that fixes the datetime issue. one thing that occurred to me is that we need to think about how we want to handle time zones. most catalogs treat time in UTC, but some catalogs like the japanese catalog and maybe the chinese catalog use local time. 

one option is to just use time-zone unaware objects at let the users deal with that all together, but this can cause issues if poeple are trying to merge data from multiple sources, say their local catalog and a finite fault inversion from ComCat.

**Comment by wsavran** [comment]
fixed with #37
